_posts/function-contracts.md

```diff
@@ -0,0 +1,410 @@
+
+Function contracts are conceptually rather simple actually. They comprise a set of conditions which characterize the behavior of the function, similar to the conditions you would use in a test case. There are different types of conditions which are also called the *clauses* of the contract. The first type of clause we will introduce here is the `ensures` clause.
+
+```rust
+#[kani::ensures(max % result == 0 && min % result == 0 && result != 0)]
+fn gcd(mut max: u64, mut min: u64) -> u64 {
+    if min > max {
+        std::mem::swap(&mut max, &mut min);
+    }
+
+    let rest = max % min;
+    if rest == 0 { min } else { gcd(min, rest) }
+}
+
+```
+
+The `ensures` clause describes the relationship of the return of the function (e.g. `result`) with the arguments that the function was called with. It is often also called a *postcondition*, because it is a *condition* that must hold after (*post*) the execution of the function. With Kani, the contents of the `ensures` clause can be any Rust expression that returns `bool`. However the expression may not perform any *side effects*, that is: allocate, deallocate or modify heap memory or perform I/O. A single function may have multiple `ensures` clauses which functions as though they had been joined with `&&`. Our example could thus also have been written as
+
+```rust
+#[kani::ensures(max % result == 0)]
+#[kani::ensures(min % result == 0)]
+#[kani::ensures(result != 0)]
+fn gcd(mut max: u64, mut min: u64) -> u64 { ... }
+```
+
+You may wonder at this point what this is actually useful for. It turns out that for a verifier the *abstraction* of the function as described by the conditions in our `ensures` clause, is much easier to reason about than running through the many recursive calls in the actual implementation of `gcd`. You may also notice that the abstraction only approximates `gcd`. `max % result == 0` and `min % result == 0` describe *a* common divisor of `max` and `min` but not necessarily the *greatest*. However in this case this is acceptable to us when it comes to verifying the callers of `gcd`. Later we will add an additional check to ensure we generate the largest divisor.
+
+Our goal will be to eventually use the efficient abstraction everywhere where `gcd` is called but first we must ensure that `gcd` respects these conditions. We need to *verify the contract*.
+
+Doing so is the same as verifying that the postcondition(s) hold for any possible input to the function. If we were to manually create a harness that does this verification it would look like this
+
+```rust
+#[kani::proof]
+fn gcd_stub_check() {
+    // First we create any possible input using a
+    // non-deterministic value
+    let max = kani::any();
+    let min = kani::any();
+
+    // We create any possible result
+    let result = gcd(max, min);
+
+    assert!(
+        // And here is our postcondition
+        max % result == 0 && min % result == 0 && result != 0
+    );
+}
+
```

Comment:
Remove extra line.

Reply:
I'm sorry these must have reappeared after a merge. I thought we'd removed all of them.